extensions/ql-vscode/src/databases/local-databases-ui.ts

```diff
@@ -978,68 +977,86 @@ export class DatabaseUI extends DisposableObject {
+      const fixedUri = await this.fixDbUri(uri);
+      // we are selecting a database folder
+      return await this.databaseManager.openDatabase(fixedUri, {
+        type: "folder",
+      });
+    } else {
+      // we are selecting a database archive or a testproj.
+      // Unzip archives (if an archive) and copy into a workspace-controlled area
+      // before importing.
+      return await this.databaseFetcher.importLocalDatabase(
+        uri.toString(true),
+        progress,
+      );
+    }
+  }
 
-      const databases: DatabaseItem[] = [];
-      const failures: string[] = [];
-      const entries = await workspace.fs.readDirectory(uri);
-      for (const entry of entries) {
-        if (entry[1] === FileType.Directory) {
-          try {
-            const fixedUri = await this.fixDbUri(Uri.joinPath(uri, entry[0]));
-            const database = await this.databaseManager.openDatabase(fixedUri, {
-              type: "folder",
-            });
-            databases.push(database);
-          } catch (e) {
-            failures.push(entry[0]);
-          }
-        }
-      }
+  /**
+   * Ask the user for a parent directory that contains all databases.
+   * Returns all valid databases, or `undefined` if the operation was canceled.
+   */
+  private async chooseDatabasesParentFolder(
+    progress: ProgressCallback,
+  ): Promise<DatabaseItem[] | undefined> {
+    const uri = await chooseDatabaseDir(true);
+    if (!uri) {
+      return undefined;
+    }
 
-      if (failures.length) {
-        void showAndLogErrorMessage(
-          this.app.logger,
-          `Failed to import ${failures.length} database(s) (${failures.join(
-            ", ",
-          )}), successfully imported ${databases.length} database(s).`,
-        );
-      } else {
-        void showAndLogInformationMessage(
-          this.app.logger,
-          `Successfully imported ${databases.length} database(s).`,
-        );
+    const databases: DatabaseItem[] = [];
+    const failures: string[] = [];
+    const entries = await workspace.fs.readDirectory(uri);
+
+    for (const [index, entry] of entries.entries()) {
+      progress({
+        step: index + 1,
+        maxStep: entries.length,
+        message: `Importing ${entry[0]}`,
+      });
+
+      if (entry[1] === FileType.Directory) {
+        try {
+          const fixedUri = await this.fixDbUri(Uri.joinPath(uri, entry[0]));
+          const database = await this.databaseManager.openDatabase(fixedUri, {
+            type: "folder",
+          });
+          databases.push(database);
+        } catch (e) {
+          failures.push(entry[0]);
+        }
       }
+    }
 
-      return databases;
+    if (failures.length) {
+      void showAndLogErrorMessage(
+        this.app.logger,
+        `Failed to import ${failures.length} database(s), successfully imported ${databases.length} database(s).`,
+        { fullMessage: `Failed folders to import:\n${failures.join("\n")}` },
```

Comment:
Include the original message in the `fullMessage`. Indentation makes it easier to read.

```suggestion
        { fullMessage: `Failed to import ${failures.length} database(s), successfully imported ${databases.length} database(s). Failed folders to import:\n  ${failures.join("\n  ")}` },
```